docs/how-to/service-dependencies.md

```diff
@@ -1,11 +1,297 @@
+
+## My web app
+
+Imagine we are running a web application, which has:
+
+- a database listening on port 3306
+- a backend server listening on port 8081, which talks to the database
+- a frontend server listening on port 8080, which talks to the backend
+
+I.E.:
+
+`frontend (8080) -> backend (8081) -> db (3306)`
+
+## Layers configuration
+
+For the above example, we may have Pebble layers configured as follows:
+
+`001-frontend.yaml`:
+
+```yaml
+services:
+  frontend-server:
+    override: replace
+    summary: frontend
+    command: python3 -m http.server 8080
+    startup: enabled
+```
+
+`002-backend.yaml`:
+
+```yaml
+services:
+  backend-server:
+    override: replace
+    summary: backend
+    command: python3 -m http.server 8081
+    startup: enabled
+```
+
+`003-database.yaml`:
+
+```yaml
+services:
+  database:
+    override: replace
+    summary: database
+    command: python3 -m http.server 3306
+    startup: enabled
+```
+
+
+```{note}
+We use Python's http module to mock the servers.
+
+Alternatively, you may choose to define those services in a single layer. For example:
+
+`001-my-webapp-layer.yaml`:
+
+```yaml
+services:
+  frontend-server:
+    override: replace
+    summary: frontend
+    command: python3 -m http.server 8080
+    startup: enabled
+  backend-server:
+    override: replace
+    summary: backend
+    command: python3 -m http.server 8081
+    startup: enabled
+  database:
+    override: replace
+    summary: database
+    command: python3 -m http.server 3306
+    startup: enabled
+```
+
+If we start the services by running:
+
+```{terminal}
+   :input: pebble run
+2024-06-28T02:17:23.347Z [pebble] Started daemon.
+[backend-server database frontend-server]
+2024-06-28T02:17:23.353Z [pebble] POST /v1/services 2.613042ms 202
+2024-06-28T02:17:23.356Z [pebble] Service "backend-server" starting: python3 -m http.server 8081
+2024-06-28T02:17:24.363Z [pebble] Service "database" starting: python3 -m http.server 3306
+2024-06-28T02:17:25.370Z [pebble] Service "frontend-server" starting: python3 -m http.server 8080
+2024-06-28T02:17:26.385Z [pebble] Started default services with change 1.
+```
+
+We can see they are all started, which can be verified by:
+
+```{terminal}
+   :input: pebble services
+Service          Startup  Current  Since
+backend-server   enabled  active   today at 10:17 CST
+database         enabled  active   today at 10:17 CST
+frontend-server  enabled  active   today at 10:17 CST
+```
+
+## Problems
+
+Apparently, the above layer configuration works but is not exactly ideal, because:
+
+- In this example, the three components kind of rely on each other, without anyone, our whole web app won't function properly.
+- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend starts, it fails to communicate with the database, so the whole web app won't function properly.
```

Comment:
```suggestion
- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend started, it would fail to communicate with the database, so the whole web app won't function properly.
```